src/EditorFeatures/TestUtilities/InProcRemoteHostClientProvider.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Options.Providers;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.Test.Utilities.Remote;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Test.Utilities.RemoteHost
+{
+    internal static class RemoteHostOptions
+    {
+        public static readonly Option2<bool> RemoteHostTest = new Option2<bool>(
+            nameof(RemoteHostOptions), nameof(RemoteHostTest), defaultValue: false);
+    }
+
+    [ExportOptionProvider, Shared]
+    internal sealed class RemoteHostOptionsProvider : IOptionProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RemoteHostOptionsProvider()
+        {
+        }
+
+        public ImmutableArray<IOption> Options { get; } = ImmutableArray.Create<IOption>(
+            RemoteHostOptions.RemoteHostTest);
+    }
+
+    internal sealed class InProcRemoteHostClientProvider : IRemoteHostClientProvider
+    {
+        [ExportWorkspaceServiceFactory(typeof(IRemoteHostClientProvider)), Shared]
+        internal sealed class Factory : IWorkspaceServiceFactory
+        {
+            [ImportingConstructor]
+            [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+            public Factory()
+            {
+            }
+
+            public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+                => new InProcRemoteHostClientProvider(workspaceServices);
+        }
+
+        private readonly HostWorkspaceServices _services;
+        private readonly AsyncLazy<RemoteHostClient?> _lazyClient;
+
+        public InProcRemoteHostClientProvider(HostWorkspaceServices services)
+        {
+            _services = services;
+
+            _lazyClient = new AsyncLazy<RemoteHostClient?>(cancellationToken =>
+            {
+                var optionService = _services.GetRequiredService<IOptionService>();
+                if (optionService.GetOption(RemoteHostOptions.RemoteHostTest))
+                {
+                    return InProcRemoteHostClient.CreateAsync(_services, runCacheCleanup: false).AsNullable();
```

Comment:
aside: what does "runCacheCleanup' even mean?

Reply:
nit: it feels like you could jsut trivially make this an async lambda and avoid teh strange .AsNullable and SpecializedTasks.Null bit.

Reply:
Not sure what runCacheCleanup is good for.